xonsh/ptk/key_bindings.py

```diff
@@ -41,3 +41,33 @@ def insert_literal_tab(event):
         """
         event.cli.current_buffer.insert_text(env.get('INDENT'))
 
+    @handle(Keys.ControlJ, filter=IsMultiline())
+    def multiline_carriage_return(event):
+        """
+        Preliminary parser to determine if 'Enter' key should send command to
+        the xonsh parser for execution or should insert a newline for continued
+        input.
+
+        Current 'triggers' for inserting a newline are:
+        - Not on first line of buffer and line is non-empty
+        - Previous character is a colon (covers if, for, etc...)
+        - User is in an open paren-block
+        - Line ends with backslash
+        - Any text exists below cursor position (relevant when editing previous
+        multiline blocks)
+        """
+        b = event.cli.current_buffer
+        if (not b.document.on_first_line and
+           not b.document.current_line.isspace()):
+            b.newline(copy_margin=True)
+        elif b.document.char_before_cursor == ':':
+            b.newline()
+            b.insert_text(env.get('INDENT'), fire_event=False)
+        elif b.document.text.count('(') > b.document.text.count(')'):
+            b.newline()
+        elif b.document.char_before_cursor == '\\':
+            b.newline()
+        elif b.document.find_next_word_beginning() is not None:
+            b.newline(copy_margin=True)
```

Comment:
Yeah -- that was one of the things besides history I didn't have working -- I've been digging through the PTK source to figure that bit out but haven't got it yet.  It *may* have to wait until we switch out `prompt` for a more customized `prompt_application` but I haven't given up on that yet.